core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/EvaluationStrategy.java

```diff
@@ -175,11 +179,25 @@ default QueryValueEvaluationStep precompile(ValueExpr arg, QueryEvaluationContex
 	}
 
 	default <T> Set<T> makeSet() {
-		return new HashSet<>();
+		return new DefaultCollectionFactory().createSet();
 	}
 
 	default <T> Queue<T> makeQueue() {
-		return new ArrayDeque<>();
+		return new DefaultCollectionFactory().createQueue();
 	}
 
+	/**
+	 * Set the collection factory that will create the collections to use during query evaluaton.
+	 *
+	 * @param a CollectionFactory that should be used during future query evaluations
+	 **/
+	@Experimental
+	public default void setCollectionFactory(Supplier<CollectionFactory> collectionFactory) {
+		// Do nothing per default. Implementations should take this value and use it
+	}
+
+	@Experimental
+	public default Supplier<CollectionFactory> getCollectionFactory() {
+		return DefaultCollectionFactory::new;
```

Comment:
Nice use of double colon operator - I had not clocked that you could use it for suppliers in this way (though makes total sense when you think about it).